File: 卡码/代码随想录/Python/动态规划/02-背包问题/01-01背包/LeetCode_0474.py
#474. 一和零
from typing import *

class Solution:
  def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
    dp:List[List[int]]=[[0]*(n+1) for x in range(m+1)]
    for str in strs:
      x,y=0,0
      for ch in str:
        if ch=="0":
          x+=1
        else:
          y+=1
      for i in range(m,x-1,-1):
        for j in range(n,y-1,-1):
          dp[i][j]=max(dp[i][j],dp[i-x][j-y]+1)
    return dp[m][n]

File: 卡码/代码随想录/Go/动态规划/02-背包问题/01-01背包/LeetCode_0474.go
// 474. 一和零
package main

import "math"

func findMaxForm(strs []string, m int, n int) int {
	dp:=make([][]int,m+1)
	for i:=0;i<m+1;i++ {
		dp[i]=make([]int,n+1)
	}
	for _,str:=range strs {
		x,y:=0,0
		for _,ch:=range str {
			if ch=='0' {
				x++
			}else{
				y++
			}
		}
		for i:=m;i>=x;i-- {
			for j:=n;j>=y;j-- {
				dp[i][j]=int(math.Max(float64(dp[i][j]),float64(dp[i-x][j-y]+1)))
			}
		}
	}
	return dp[m][n]
}
